### PR TITLE
Optimizing Creation of MDB entities

### PR DESCRIPTION
**Pull request by anandjakhaniya**
**Ticket:** https://vertaai.atlassian.net/browse/VR-4297

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java [hunk 1]
```diff
@@ -93,63 +93,63 @@ public DatasetDAORdbImpl(AuthService authService, RoleService roleService) {
...
+        dataset.getWorkspaceType(),
+        LOGGER);
   }
 
   @Override
   public Dataset createDataset(Dataset dataset, UserInfo userInfo)
       throws InvalidProtocolBufferException {
-    // Check entity already exists
-    checkDatasetAlreadyExist(dataset);
+    createRolesForDataset(dataset, userInfo);
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: What happens if it already exists? Will the role service return an error? Or maybe add duplicated entries? I'd recommend still performing the check first. You can use a separate session for that, right?

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java [hunk 2]
```diff
@@ -629,8 +629,9 @@ public Dataset updateDatasetName(String datasetId, String datasetName)
               .setWorkspaceTypeValue(datasetObj.getWorkspace_type())
               .build();
       // Check entity already exists
-      checkDatasetAlreadyExist(dataset);
+      checkDatasetAlreadyExist(session, dataset);
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: I think it's fine to keep this within the transaction. It will help ensure consistency. We should just remove rpc from within the transactions.

File: backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java [hunk 3]
```diff
@@ -157,6 +157,8 @@ public DatasetVersion createDatasetVersion(
         DatasetVersionEntity datasetVersionEntity =
             RdbmsUtils.generateDatasetVersionEntity(datasetVersion);
         session.save(datasetVersionEntity);
+        setDatasetUpdateTime(session, Collections.singletonList(datasetVersion.getDatasetId()));
+        transaction.commit();
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java:145]: Always create the role binding before creating the object, and always delete the object before deleting the bindings. We can have the system giving permission for an object that doesn't exist, but we can't have the system having an object with no permission.

File: backend/src/main/java/ai/verta/modeldb/experiment/ExperimentDAORdbImpl.java [hunk 4]
```diff
@@ -528,31 +531,6 @@ public Experiment deleteExperimentAttributes(
     }
   }
 
-  @Override
-  public Boolean deleteExperiment(String experimentId) {
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experiment/ExperimentDAORdbImpl.java]: Where did this go?

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java [hunk 5]
```diff
@@ -393,51 +396,6 @@ public ExperimentRun insertExperimentRun(ExperimentRun experimentRun, UserInfo u
     return hyrParamMappings;
   }
 
-  @Override
-  public Boolean deleteExperimentRun(String experimentRunId) {
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java]: Where did this go?

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java [hunk 6]
```diff
@@ -282,8 +281,9 @@ public CommitEntity getCommitEntity(
         throw new ModelDBException("Commit is associated with Label", Code.FAILED_PRECONDITION);
       }
 
+      session.beginTransaction();
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: I believe it's fine to leave the logic above inside the transaction. It should be pretty fast. We can revisit if not.

**Review by conradoverta — CHANGES_REQUESTED** [review]
Be careful not to get greedy removing things from transactions and ending up with a bug. We should just remove rpcs right now. Create role bindings before creating the object, plus delete the object before deleting role bindings.

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java (hunk 1 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: this is moving into the below single transaction. Role service update do nothing if role binding is already added but I will confirm it again.

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: `checkDatasetAlreadyExist(dataset);` is calling from transaction scope so it will lead the transaction under transaction so I have make it with a single transaction.

File: backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java (hunk 3 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java:145]: @conradoverta  `Always create the role binding before creating the object,` is not possible for all entities because for the project we are creating ID our self it will be fine to create before creating it but like repository entity, we can not do it before creating because repository id is auto genearted by the Hibernate so we got id after entity creation. For role binding, we need to add entity id so in that case, we should create role binding after entity creation.

File: backend/src/main/java/ai/verta/modeldb/experiment/ExperimentDAORdbImpl.java (hunk 4 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experiment/ExperimentDAORdbImpl.java]: Not in use so removed it. we have another function called `deleteExperiments` to delete entities.

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java (hunk 5 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java]: Not in use so removed it. we have another function called `deleteExperimentRuns` to delete entities.

File: backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java (hunk 3 above)
Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java:145]: That will cause issues. If we don't satisfy the rule I wrote, the code will have a bug waiting to happen. We'll have to figure out how to resolve it.

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java (hunk 1 above)
Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: Please do and let me know here.

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java (hunk 2 above)
Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: Please add a TODO to remove the excess of transactions. Not a big deal here since it's not a very frequent path.

File: backend/src/main/java/ai/verta/modeldb/experiment/ExperimentDAORdbImpl.java (hunk 4 above)
Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/experiment/ExperimentDAORdbImpl.java]: Make a separate PR for it. Let's not shove unrelated changes together for no reason.

**Review by conradoverta — COMMENTED** [review]
Minor things. I think we can close this out soon.

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java (hunk 1 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: @conradoverta It is created a new role binding with a new dataset id so I am reverting my change and add exists check before the role creation.

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: Reverted changes (`remove the excess of transactions`)

File: backend/src/main/java/ai/verta/modeldb/experiment/ExperimentDAORdbImpl.java (hunk 4 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experiment/ExperimentDAORdbImpl.java]: reverted

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java (hunk 5 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java]: reverted

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java (hunk 6 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: reverted as per your comment

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java [hunk 7]
```diff
@@ -114,42 +114,44 @@ public Dataset createDataset(Dataset dataset, UserInfo userInfo)
...
-          dataset.getDatasetVisibility());
-
       transaction.commit();
       LOGGER.debug("Dataset created successfully");
       TelemetryUtils.insertModelDBDeploymentInfo();
       return datasetEntity.getProtoObject();
     }
   }
 
+  private void createRolesForDataset(Dataset dataset, UserInfo userInfo) {
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: createRoleBindingsForDataset  or createBindingsForDataset would be a better name, right?

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java [hunk 8]
```diff
@@ -345,9 +347,10 @@ public Boolean deleteDatasets(List<String> datasetIds) throws InvalidProtocolBuf
       for (DatasetEntity datasetObj : datasetEntities) {
         session.delete(datasetObj);
       }
+      transaction.commit();
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: this is better suited in other batch of PRs related to delete?

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java [hunk 9]
```diff
@@ -617,6 +619,7 @@ public Dataset updateDatasetName(String datasetId, String datasetName)
       // Check entity already exists
       checkDatasetAlreadyExist(dataset);
 
+      Transaction transaction = session.beginTransaction();
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: might as well move it two lines below.

Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: wouldn't a similar change be required for updateDatasetDescription and updateDatasetAttributes as well?

File: backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java (hunk 3 above)
Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java:145]: we can create the id here and pass it to be used at https://github.com/VertaAI/modeldb/blob/master/backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAO.java#L330

File: backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java [hunk 10]
```diff
@@ -670,8 +672,8 @@ private void deleteExperimentsWithPagination(Session session, List<String> proje
           roleBindingNames.add(ownerRoleBindingName);
         }
       }
-      roleService.deleteRoleBindings(roleBindingNames);
       transaction.commit();
+      roleService.deleteRoleBindings(roleBindingNames);
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java:676]: these delete optimizations can be addressed in the PR for deleting entities.

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAO.java [hunk 11]
```diff
@@ -49,6 +49,8 @@ BranchEntity getBranchEntity(Session session, Long repoId, String branchName)
 
   void deleteBranchByCommit(Long repoId, String commitHash);
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAO.java]: it seems this can be removed

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java [hunk 12]
```diff
@@ -335,10 +316,35 @@ public RepositoryEntity getRepositoryById(Session session, RepositoryIdentificat
             session, commitEntity.getCommit_hash(), ModelDBConstants.MASTER_BRANCH, repository);
       }
       session.getTransaction().commit();
+      if (create) {
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java:319]: 1. once https://github.com/VertaAI/modeldb/pull/649 is in we should create the repo with delete flag true.
1. create the role binding
    1. if the role binding creation succeeds, update the delete flag
    1.if the role binding creation fails, return error.

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java [hunk 13]
```diff
@@ -392,6 +397,12 @@ private void deleteRoleBindingsOfAccessibleResources(List<RepositoryEntity> allo
       ListBranchesRequest.Response listBranchesResponse =
           listBranches(
               ListBranchesRequest.newBuilder().setRepositoryId(request.getRepositoryId()).build());
+
+      CommitPaginationDTO commitPaginationDTO =
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java]: I prefer the delete optimizations to be part of different PR

**Review by ravishetye — CHANGES_REQUESTED** [review]

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java (hunk 2 above)
Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: along with TODO , please file a jira referring the TODO

Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: Jira Ticket: https://vertaai.atlassian.net/browse/VR-4405

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java (hunk 12 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java:319]: I will update this logic after merge #649 .

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java (hunk 7 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: renamed

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java (hunk 9 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: moved it to all update* functions

File: backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java (hunk 10 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java:676]: reverted

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAO.java (hunk 11 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAO.java]: removed

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java (hunk 13 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java]: reverted

File: backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java (hunk 8 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/dataset/DatasetDAORdbImpl.java]: reverted

File: backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java (hunk 3 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java:145]: Modified changes

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java (hunk 12 above)
Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java:319]: Please file a ticket so that we can track it, refer this comment so we have context

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java [hunk 14]
```diff
@@ -703,22 +709,16 @@ public BranchEntity getBranchEntity(Session session, Long repoId, String branchN
...
-              .append(BranchEntity.class.getSimpleName())
-              .append(" br where br.id.repository_id = :repositoryId ")
-              .append(" AND br.commit_hash = :commitHash ");
-      Query deleteBranchQuery = session.createQuery(deleteBranchesHQLBuilder.toString());
-      deleteBranchQuery.setParameter("repositoryId", repoId);
-      deleteBranchQuery.setParameter("commitHash", commitHash);
-      deleteBranchQuery.executeUpdate();
-      session.getTransaction().commit();
-    }
+  public void deleteBranchByCommit(Session session, Long repoId, String commitHash) {
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java:712]: can we move this to other PR as well?

**Review by ravishetye — APPROVED** [review]
LGTM

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java (hunk 12 above)
Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java:319]: https://vertaai.atlassian.net/browse/VR-4406

File: backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java [hunk 15]
```diff
@@ -122,12 +121,12 @@ public DatasetVersion createDatasetVersion(
           getDatasetVersionFromRequest(authService, request, userInfo, existingDatasetVersion);
 
       if (datasetVersionList.size() == 1) {
-        transaction.commit();
-        return datasetVersionList.get(0);
+        DatasetVersion datasetVersion = datasetVersionList.get(0);
+        createRoleBindingsForDatasetVersion(userInfo, datasetVersion);
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java]: when datasetVersionList size is 1 it means the new version was similar to the old version, in which case we need not create the bindings as was happening in older code.

**Review by ravishetye — CHANGES_REQUESTED** [review]

File: backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java (hunk 15 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java]: Remove RoleBinding creation step

**Review by ravishetye — APPROVED** [review]
